### PR TITLE
rgw: when resharding store progress json

### PR DESCRIPTION
**Pull request by mkogan1**
only when both verbose and out are specified.

Fixes: https://tracker.ceph.com/issues/42051

Signed-off-by: Mark Kogan <mkogan@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/rgw/rgw_reshard.cc [hunk 1]
```diff
@@ -529,14 +529,14 @@ int RGWBucketReshard::do_reshard(int num_shards,
 
   verbose = verbose && (formatter != nullptr);
 
-  if (verbose) {
+  if (verbose && out) {
```
Comment (cbodley) [inline src/rgw/rgw_reshard.cc]: would it work to just change the verbose assignment above?
```diff
-  verbose = verbose && (formatter != nullptr);
+  verbose = verbose && (formatter != nullptr) && (out != nullptr);
```

File: src/rgw/rgw_reshard.cc [hunk 2]
```diff
@@ -604,23 +604,19 @@ int RGWBucketReshard::do_reshard(int num_shards,
 	  }
 	}
 
-	if (verbose) {
+	if (verbose && out) {
 	  formatter->close_section();
-	  if (out) {
-	    formatter->flush(*out);
-	  }
+	  formatter->flush(*out);
```
Comment (mattbenjamin) [inline src/rgw/rgw_reshard.cc]: ack

File: src/rgw/rgw_reshard.cc [hunk 3]
```diff
@@ -604,23 +604,19 @@ int RGWBucketReshard::do_reshard(int num_shards,
...
     }
   }
 
-  if (verbose) {
+  if (verbose && out) {
     formatter->close_section();
-    if (out) {
-      formatter->flush(*out);
-    }
+    formatter->flush(*out);
```
Comment (mattbenjamin) [inline src/rgw/rgw_reshard.cc]: ack

File: src/rgw/rgw_reshard.cc [hunk 4]
```diff
@@ -993,8 +989,11 @@ int RGWReshard::process_single_logshard(int logshard_num)
 	}
 
 	RGWBucketReshard br(store, bucket_info, attrs, nullptr);
-	ret = br.execute(entry.new_num_shards, max_entries, false, nullptr,
-			 nullptr, this);
+        // ret = br.execute(entry.new_num_shards, max_entries, false, nullptr,
```
Comment (mattbenjamin) [inline src/rgw/rgw_reshard.cc]: should not leave commented code

File: src/rgw/rgw_reshard.cc (hunk 1 above)
Comment (mattbenjamin) [reply src/rgw/rgw_reshard.cc]: I don't particularly like using "verbose" as an alias for "have formatter"

File: src/rgw/rgw_reshard.cc [hunk 5]
```diff
@@ -993,8 +989,11 @@ int RGWReshard::process_single_logshard(int logshard_num)
...
+	// 		 nullptr, this);
+        Formatter* formatter = new JSONFormatter(false);
```
Comment (yuvalif) [inline src/rgw/rgw_reshard.cc]: i think this is a leak. you should probably use unique_ptr

Comment (mkogan1) [reply src/rgw/rgw_reshard.cc]: Removed & apology, this hunk was committed by mistake,
(it was changed for testing purpose only).
I did not intend to revert commit cc27e353cb821dcba6f1a484bf2ea65a0c109778

File: src/rgw/rgw_reshard.cc (hunk 1 above)
Comment (mkogan1) [reply src/rgw/rgw_reshard.cc]: Yes, changed to:
https://github.com/ceph/ceph/blob/4221452e90e01d5f27cf3ccf92d0b3a359218c30/src/rgw/rgw_reshard.cc#L530

File: src/rgw/rgw_reshard.cc (hunk 4 above)
Comment (mkogan1) [reply src/rgw/rgw_reshard.cc]: Removed & apology, this hunk was committed by mistake,
(it was changed for testing purpose only).
I did not intend to revert commit cc27e35

File: src/rgw/rgw_reshard.cc (hunk 2 above)
Comment (mkogan1) [reply src/rgw/rgw_reshard.cc]: Reverted & apologizing for the confusion, this hunk got staged by mistake, did not mean to add a formatter / revert commit cc27e35  
Just was reproducing the memory leak as in nautilus,mimic,luminous
For example in Nautilus it was:
https://github.com/ceph/ceph/blob/534c2f14693439413791049826ca23d1706c7a8d/src/rgw/rgw_reshard.cc#L1023-L1024

**Pull request merged**